cpp/README.md

```diff
@@ -365,3 +365,60 @@ add_executable(...)
 
 bob_end_package()
 ```
+
+#### Embedded third parties
+
+External libraries required to build or test your C++ project can be either
+directly added to the git repository or as a git submodule. The standard
+root location for this kind of files is the `3rdparty/` directory but can be
+overriden with the `${PROJECT_NAME}_3RDPARTY_DIR` CMake variable.
+
+Adding single-file/header-only C++ libraries directly to the git repository
+of your project is acceptable in general, like catch2 or the JSON library
+of Niels Lohmann for instance.
+
+More significant dependencies should be considered as pure external
+dependencies. But it can also be very convenient to have them as git
+submodules, and be able to switch between the two.
+
+This project provides helper functions to deal with these dependencies:
+
+###### cpp_cc_git_submodule
+
+````cmake
+#
+# cpp_cc_git_submodule(source_dir
+#                      [DISABLED]
+#                      SUBDIR path
+#                      [BUILD] [<arguments>]
+#                      [PACKAGE] [<arguments>]
+#                      GIT_ARGS [<arguments>])
+#
+# Add a CMake option in the cache to control whether the
+# submodule is used or not (default ON). The option is named after the source
+# directory passed in first argument, for instance:
+#   cpp_cc_git_submodule(src/eigen)
+# adds the following CMake cached option:
+#  ${PROJECT_NAME}_3RDPARTY_USE_SRC_EIGEN:BOOL=ON
+#
+# If enabled, then the submodule is fetched if missing in the working copy.
+#
+# If the DISABLED argument is provided, then the default value for the CMake
+# option is OFF.
+#
+# If the BUILD argument is provided then the directory is added to the build
+# through the add_subdirectory CMake function. Arguments following the BUILD
+# arguments are passed to the add_subdirectory function call.
+#
+# The optional SUBDIR argument is used by the BUILD argument to determine
+# the path to the directory added to the build. The path specified is relative
+# to the path to the git submodule.
+#
+# If the PACKAGE argument is provided and the CMake option to determine whether
+# the git submodule should be used or not is FALSE, then a call to the find_package
+# function is made with the arguments specified to the PACKAGE option.
+#
+# Default options passed to the `git submodule update` command are `--init --recursive`.
+# If the opt_GIT_ARGS argument is provided, then its value supersedes the default options.
```

Comment:
```suggestion
# If the GIT_ARGS argument is provided, then its value supersedes the default options.
```
?

Also, would it make sense to have a shallow clone and state that explicitly?

Reply:
I agree with you @matz-e 
I pushed 70d3b7a which does a shallow clone by default, which can be overridden through GIT_ARGS.

Reply:
Are you ok with that @pramodk and @matz-e  ?

Reply:
If there is any hpc coding convention specific options, better to have prefix?

Reply:
No, this is a function parameter, not a CMake variable in the global scope:

```
cpp_cc_git_submodule(pybind11 BUILD GIT_ARGS --init --depth 42)
```

Reply:
Ah ok! 👍